Add support for remote notifications

aaemnnosttv opened the pull request:
## Summary

Addresses issue #1110

## Relevant technical choices
- Proxy API responses are mocked during E2E
- Added `dompurify` to allow for limited HTML in content passed to notification `description`
- Increased bundle size limit by 10KB (see #1180)
- ~~Updated VRT due to visual changes (see https://github.com/google/site-kit-wp/pull/1170#issuecomment-591315512)~~

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

aaemnnosttv commented on includes/Core/Notifications/Notifications.php:
```diff
@@ -0,0 +1,262 @@
+										'site_id'         => $credentials['oauth2_client_id'],
+										'site_secret'     => $credentials['oauth2_client_secret'],
+										'notification_id' => $data['notificationID'],
+										'notification_state' => $data['notificationState'],
```
@felixarntz I'm not sure why, but PHPCS says that this alignment is correct.

tofumatt replied:
I think it wants to `=>` to be aligned?

```php
'site_id'            => $credentials['oauth2_client_id'],
'site_secret'        => $credentials['oauth2_client_secret'],
'notification_id'    => $data['notificationID'],
'notification_state' => $data['notificationState'],
```

aaemnnosttv replied:
It should want that, but it automatically fixes it as highlighted above (which should be wrong)

aaemnnosttv commented on includes/Core/Notifications/Notifications.php:
```diff
@@ -0,0 +1,262 @@
+
+		if ( ! empty( $decoded['error'] ) ) {
+			throw new Exception( $decoded['error'] );
+		}
```
@felixarntz Should we ever display an error message returned by the remote? Currently it's logged in the console only.

tofumatt replied:
A generic error message (like a 500, if the proxy's notification endpoint was down/erroring) seems wrong to display, because it's not very actionable/helpful to the user. I'd think an allow-list of errors to display would work if wanted, but otherwise it seems best to log them to the console for debugging/support. Otherwise hide 'em?

felixarntz replied:
Yeah I think we should stick to only logging and don't render any user-facing messages. It should just silently error then.

aaemnnosttv commented on includes/Core/Notifications/Notifications.php:
```diff
@@ -0,0 +1,262 @@
+			sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), $param ),
+			array( 'status' => 400 )
+		);
+	}
```
@felixarntz I've thought about refactoring with something like this before since there is quite a bit of duplication for this kind of thing in the datapoint handlers. Ideally we would use core's `args` definitions and leverage the native validation the API already provides. The problem is that our data requests stuff all the parameters into one `data` parameter (except for a few outliers like module activation), and the REST API doesn't support nested validation except for collections. This may not be the best place to discuss this but just thought I'd raise it.

felixarntz replied:
Yeah I thought about this initially too, but our support for batch requests would make use of that core feature challenging. I think for now we should keep this internal to the respective handlers (although adding utility methods, potentially even a validation class or trait for common validation features is certainly worth considering).

aaemnnosttv replied:
> Yeah I thought about this initially too, but our support for batch requests would make use of that core feature challenging.

It would require that we change the way batch requests are handled to dispatch the individual REST requests (internally-w/o HTTP) in the batch handler rather than the way it is now but I think it's possible. Better to discuss this in a new issue though.

felixarntz commented on tests/phpunit/integration/Core/REST_API/REST_RoutesTest.php:
```diff
@@ -53,6 +53,8 @@ public function test_register() {
 			'/' . REST_Routes::REST_ROOT . '/core/site/data/setup-tag',
 			'/' . REST_Routes::REST_ROOT . '/core/site/data/connection',
+			'/' . REST_Routes::REST_ROOT . '/core/site/data/notifications',
+			'/' . REST_Routes::REST_ROOT . '/core/site/data/mark-notification',
```
Yeah that makes sense. Could you do that in a follow-up PR? If we can get it into this release, we can even just connect it to the same issue. I just don't want that work to block this PR.

felixarntz commented:
Current appearance of the notification:
<img width="927" alt="Screenshot 2020-02-25 at 14 19 08" src="https://user-images.githubusercontent.com/3531426/75293874-fcdcbb00-57db-11ea-81da-831bbf154d35.png">

felixarntz commented:
@aaemnnosttv Could you maybe look into how we could support some basic inline HTML tags in the `content` returned? Would be useful to support at least a few whitelisted things such as `strong`, `em`, `br`, `a href`. It would be great if we could include that in this initial implementation, particularly because we would want to break the notification text into two lines with `br`.

felixarntz commented on includes/Core/Notifications/Notification.php:
```diff
@@ -0,0 +1,110 @@
+	 */
+	public function prepare_for_js() {
+		return array(
+			'slug'           => $this->get_slug(),
```
@aaemnnosttv That was not originally the plan, but maybe we should change this to `id` (just here, for JS)? "slug" is kind of the term only used in PHP, for consistency maybe `id` would be better for JS.

aaemnnosttv replied:
Sure. I assumed it should be included since we need it for identifying the notice in the click handlers. I'll update the key to `id`.

kevinpdicarlo commented:
how come i am getting all theses threads??

Thank You,
Circle 22

Harmonize Humanity




On Wed, Feb 26, 2020 at 1:28 AM Evan Mattson <notifications@github.com>
wrote:

> *@aaemnnosttv* commented on this pull request.
> ------------------------------
>
> In includes/Core/Notifications/Notification.php
> <https://github.com/google/site-kit-wp/pull/1170#discussion_r384295753>:
>
> > +	 * @return string Unique notification slug.
> +	 */
> +	public function get_slug() {
> +		return $this->slug;
> +	}
> +
> +	/**
> +	 * Prepares the JS representation of the Notification.
> +	 *
> +	 * @since n.e.x.t
> +	 *
> +	 * @return array
> +	 */
> +	public function prepare_for_js() {
> +		return array(
> +			'slug'           => $this->get_slug(),
>
> Sure. I assumed it should be included since we need it for identifying the
> notice in the click handlers. I'll update the key to id.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/site-kit-wp/pull/1170?email_source=notifications&email_token=AKV4TNVETFYBJ27AKJJB3IDREYD2FA5CNFSM4K3G3YH2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCW6BOHA#discussion_r384295753>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKV4TNS6NAEXJB5Z3Q6MDLLREYD2FANCNFSM4K3G3YHQ>
> .
>

aaemnnosttv commented:
@kevinpdicarlo - looks like you're [watching/following](https://github.com/google/site-kit-wp/watchers) the repository.

See here about updating what you get notified about:
https://help.github.com/en/github/receiving-notifications-about-activity-on-github/watching-and-unwatching-repositories

TL;DR, you're probably looking for "Not Watching" or "Releases Only" 😄 
![image](https://user-images.githubusercontent.com/1621608/75324711-35ce6b80-5880-11ea-846d-9913db36a0ea.png)

aaemnnosttv commented:
@felixarntz all of your feedback should be addressed now.

Regarding E2E, I updated the API mock to intercept requests to the proxy and return mock responses instead of overriding (and re-implementing) our core routes like we do in others. This makes it a bit more end-to-end and actually ensures that we are passing the necessary parameters to the endpoint as well. This is the better way to do it where possible I think.

As for the support for HTML in notification descriptions, I've added `dompurify` and allowed for the select elements and attributes you mentioned. One byproduct of this change is that we needed to separate the description into its own element (before it was sharing a `<p>`). This resulted in a few breaking changes to VRT but I think they are actually an improvement so I've gone ahead and approved them but you can see the full changes here.
[vrt-report.zip.txt](https://github.com/google/site-kit-wp/files/4254466/vrt-report.zip.txt)
_GH doesn't allow uploading zip files so just remove the `.txt` after downloading 😄_ 

![image](https://user-images.githubusercontent.com/1621608/75329020-9a40f900-5887-11ea-93db-28bec27ed5ec.png)

![image](https://user-images.githubusercontent.com/1621608/75328945-7978a380-5887-11ea-93af-d010d4d36b20.png)

aaemnnosttv commented:
I had to bump the bundlesize limit slightly to allow for dompurify. I'll open a new issue about addressing the increasing size.

tofumatt commented:
@googlebot I consent.

tofumatt commented on assets/js/components/notifications/dashboard-core-site-alerts.js:
```diff
@@ -0,0 +1,88 @@
+	async componentDidMount() {
+		const notifications = await getNotifications().catch( ( e ) => {
+			// eslint-disable-next-line no-console
+			console.warn( 'Error caught while fetching notifications', JSON.stringify( e ) );
```
I often use `global.console` for these. It probably doesn't matter? I dunno!

tofumatt replied:
Also: why stringify the error? `console` can output objects.

aaemnnosttv replied:
Yeah, I know but it doesn't work when logging from E2E tests.

https://travis-ci.com/google/site-kit-wp/jobs/291425794#L1104

I didn't mean to keep it like that though so I'll update it.

tofumatt commented on assets/js/components/notifications/dashboard-core-site-alerts.js:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	async componentDidMount() {
+		const notifications = await getNotifications().catch( ( e ) => {
```
For `async/await` code it's usually the norm to do try/catch blocks instead of `catch`. Any reason to prefer catch here?
```suggestion
		try {
			const notifications = await getNotifications();
			this.setState( { notifications } );
		} catch ( err ) {
			// eslint-disable-next-line no-console
			console.warn( 'Error caught while fetching notifications', JSON.stringify( err ) );
		}
```

I dunno what `notifications` get set to when this fails, but it seems off to set notifications if there was an error fetching them too.

tofumatt commented on assets/js/components/notifications/dashboard-core-site-alerts.js:
```diff
@@ -0,0 +1,88 @@
+		}
+
+		return notifications.map(
+			( notification ) => <Notification
```
Just a nitpick, but I find this a bit tougher to read than:

```js
( notification ) => (
  <Notification
    {…props}
  />
)
```

tofumatt commented on assets/js/components/notifications/dashboard-core-site-alerts.js:
```diff
@@ -0,0 +1,88 @@
+				ctaLink={ notification.ctaURL || '' }
+				ctaLabel={ notification.ctaLabel || '' }
+				ctaTarget={ notification.ctaTarget || '' }
+				dismiss={ notification.dismissLabel || __( 'OK, Got it!', 'google-site-kit' ) }
```
```suggestion
				dismiss={ notification.dismissLabel || __( 'OK, got it!', 'google-site-kit' ) }
```

The "G" here shouldn't be capitalised if not a new sentence.

aaemnnosttv replied:
There's a bunch of these in the codebase. I've updated them all so let's see how that goes...

tofumatt commented on includes/Core/Notifications/Notification.php:
```diff
@@ -0,0 +1,110 @@
+		$this->args = array_merge(
+			array(
+				'title'            => '',
+				'content'          => '',
```
Since these are always empty-string by default, any reason the component does `notification.content || ''`  when passing props? Seems unneeded. 🤷‍♂ Belt and braces I guess?

aaemnnosttv replied:
This ensures that the keys are always set on the PHP side.

aaemnnosttv commented:
> Can you clarify why this is necessary? Wouldn't it be possible to use `Fragment dangerouslySetInnerHTML=...` instead? It would be great if we could avoid that change, so that it doesn't result in additional paragraphs just for the learn more link.

@felixarntz `Fragment`s don't support using `dangerouslySetInnerHTML` 😞 
See https://github.com/facebook/react/issues/12014

> It would be great if we could avoid that change, so that it doesn't result in additional paragraphs just for the learn more link.

I think the UI is a bit cleaner and more consistent with the separation myself, but if we wanted to keep them, we could wrap `description` and the learn more with `span`s instead and then wrap that with a `<p>` tag to let them render inline.

> These visual changes seem off to me:

@tofumatt They are _different_, but not unexpected. Like I said, I think they're an improvement but we can revert that if necessary.
See https://github.com/google/site-kit-wp/pull/1170#issuecomment-591315512

felixarntz commented:
@aaemnnosttv 
> I think the UI is a bit cleaner and more consistent with the separation myself, but if we wanted to keep them, we could wrap `description` and the learn more with `span`s instead and then wrap that with a `<p>` tag to let them render inline.

Yeah I thought about that too, let's use `span`s then and maintain the original appearance.

felixarntz commented on assets/js/components/notifications/dashboard-core-site-alerts.js:
```diff
@@ -0,0 +1,91 @@
+					dismiss={ notification.dismissLabel || __( 'OK, Got it!', 'google-site-kit' ) }
+					isDismissable={ notification.dismissible }
+					onCTAClick={ async () => {
+						await acceptNotification( notification.id );
```
We need to make the notice dismiss here, also when the user accepts it. It could either be through manual logic in this callback here, or alternatively we could introduce some kind of `dismissOnCTA` flag prop in `Notification` to generally support that?

felixarntz replied:
The thing is right now, it opens the link in a new tab (which is what should happen). The problem is then the notice still remains open in the other window. We need to make sure it dismissed on the JS side.

aaemnnosttv replied:
I just pushed a small update to address this although it needs testing still.

Basically, if no `onCTAClick` handler is passed, the notification will dismiss on CTA click if `dismissible`.

If an `onCTAClick` handler _is passed_ then that callback is responsible for calling `dismiss` or not, which is passed to the callback as the second argument (first being the event). If the event isn't dismissible, the `dismiss` function is a no-op.

felixarntz commented on assets/js/components/notifications/notification.js:
```diff
@@ -74,6 +89,19 @@ class Notification extends Component {
+			onCTAClick( e, dismiss );
+		} else {
+			dismiss( e );
+		}
```
Revisiting this, wouldn't it make more sense to make `handleCTAClick` `async` and then `await` `onCTAClick`, without passing `dismiss` to it? Then we would always call `dismiss` at the end. This would make it more similar to `handleDismiss` and avoid the extra parameter.

aaemnnosttv replied:
Sure we could do that. My initial approach here assumes that there might be cases where we don't want to dismiss on CTA click, even if the notification is dismissible, but that does sound like a bit of an edge case 😄 

This looks cleaner indeed though so assuming the above isn't a concern, let's go with it 👍

felixarntz commented on assets/js/components/notifications/notification.js:
```diff
@@ -55,14 +60,24 @@ class Notification extends Component {
-		const card = this.cardRef.current;
-
+	async handleDismiss( e ) {
+		e.persist();
```
What's this for?

aaemnnosttv replied:
Since these events are actually [`SyntheticEvents`](https://reactjs.org/docs/events.html#overview) from React, we need to persist them using this method to ensure they are not disposed of to make them available to any subsequent callbacks. We don't actually use the events right now, but if we're going to pass it, it should be persisted.

aaemnnosttv commented:
@felixarntz I've simplified the `handleCTAClick` as you suggested and testing has been going well.
Ready for another pass 👍
